feathr-impl/src/main/scala/com/linkedin/feathr/offline/util/FeathrUtils.scala

```diff
@@ -128,4 +176,61 @@ private[offline] object FeathrUtils {
     FeathrUtils.getFeathrJobParam(sparkConf, FeathrUtils.ENABLE_DEBUG_OUTPUT).toBoolean
   }
 
+  /**
+   * Dump a dataframe to disk if debug mode is enabled.
+   * The function will check the features against the 'spark.feathr.debug.feature.names',
+   * making sure there's overlap before dumping the dataframe.
+   * @param ss spark session
+   * @param df input dataframe
+   * @param features features in the input dataframe
+   * @param tag tag used in the log
+   * @param pathSuffix path suffix used to dump the dataframe
+   */
+  def dumpDebugInfo(ss: SparkSession,
+                            df: DataFrame,
+                            features: Set[String],
+                            tag: String,
+                            pathSuffix: String): Unit = {
+    if (isDebugMode(ss)) {
+      val basePath = FeathrUtils.getFeathrJobParam(ss.sparkContext.getConf, FeathrUtils.DEBUG_OUTPUT_PATH)
+      val debugFeatureNames = FeathrUtils.getFeathrJobParam(ss.sparkContext.getConf, FeathrUtils.DEBUG_FEATURE_NAMES)
+        .split(FeathrUtils.STRING_PARAMETER_DELIMITER).toSet
+      val featureNames = "features_" + features.mkString("_") + "_"
+      if (debugFeatureNames.isEmpty || features.intersect(debugFeatureNames).nonEmpty) {
+        val savePath = SimplePath(basePath + "/" + featureNames + pathSuffix)
+        log.info(s"${tag}, Start dumping data ${featureNames} to ${savePath}")
+        if (!df.isEmpty) {
+          SparkIOUtils.writeDataFrame(df, savePath, Map(), List())
```

Comment:
With the large number of dumpDebugInfo calls, is there a concern that this will result in a really long run time and a huge file which is not really readable in the case of large dfs? Do we want to consider limiting the size of the df that gets written? If not then we should document the behavior of the debug mode clearly that the whole DF will get written to disk which can take long and be unreadable for large DFs.

Reply:
This is only for debug mode, the logging is turned off by default. We check the feature names in the dataframe so only the those specified in 'spark.feathr.debug.feature.names' will be dumped.

I tried to provide a filter before dump, however, the config would be non-trivial. It is essentially another config DSL, e.g. need to specify keys (may have multiple key columns) for each feature, observation at different steps. This is an overkill.

Dumping the some dataframe as a whole and using notebook to do analysis when needed is a better solution to me.